sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1612 @@
+buffer accessors to existing nodes in the graph are captured as an edge. Using
+`handler::depends_on()` will also create a graph edge when passed an event
+returned from a queue submission captured by a queue recording to the same graph.
+|===
+
+==== Queue Recording API
+
+When using the record & replay API to construct a graph by recording a queue,
+nodes and edges are captured as follows.
+
+Table {counter: tableNumber}. Recorded Graph Definition.
+[%header,cols="1,3"]
+|===
+| Concept | Description
+
+| Node
+| A node in a queue recorded graph represents a command group submission to the
+device associated with the queue begin recorded. Each submission encompasses
+either one or both of a.) some data movement, b.) a single asynchronous kernel
+launch. Nodes cannot define forward edges, only backwards. That is, kernels can
+only create dependencies on command-groups that have already been submitted.
+This means that transparently a node can depend on a previously recorded graph
+(sub-graph), which works by creating edges to the individual nodes in the old
+graph. Explicit memory operations without kernels, such as a memory copy, are
+still classed as nodes under this definition, as the
+{explicit-memory-ops}[SYCL 2020 specification states] that these can be seen as
+specialized kernels executing on the device.
+
+| Edge
+| An edge in a queue recorded graph is expressed through command group
+dependencies in one of two ways. Firstly, through buffer accessors that
+represent data dependencies between two command groups captured as nodes.
+Secondly, by using the `handler::depends_on()` mechanism inside a command group
+captured as a node. However, for an event passed to `handler::depends_on()` to
+create an edge, it must be an event returned from a queue
+submission captured by the same graph. Otherwise, a synchronous error will be
+thrown with error code `invalid`. `handler::depends_on()` can be
+used to express edges when a user is working with USM memory rather than SYCL
+buffers.
+|===
+
+==== Sub-Graph
+
+A node in a graph can take the form of a nested sub-graph. This occurs when
+a command-group submission that invokes `handler::ext_oneapi_graph()` with an
+executable graph object is added to the graph as a node.
+
+=== API Modifications
+
+[source, c++]
+----
+namespace sycl {
+namespace ext::oneapi::experimental {
+
+// State of a queue, returned by queue::ext_oneapi_get_state()
+enum class queue_state {
+  executing,
+  recording
+};
+
+namespace property {
+
+namespace graph {
+
+class no_cycle_check {
+  public:
+    no_cycle_check() = default;
+};
+
+} // namespace graph
+
+namespace node {
+
+class depends_on {
+  public:
+    template<typename... NodeTN>
+    depends_on(NodeTN... nodes);
+};
+
+} // namespace node
+} // namespace property
+
+// Device query for level of support
+namespace info {
+namespace device {
+struct graphs_support;
+
+enum class graph_support_level {
+  unsupported = 0,
```

Comment:
I was thinking it would give the users the convenience of writing
```cpp
if (!device.get_info<info::device::graph_support>)
```
Rather than:
```cpp
if (graph_support_level::unsupported == device.get_info<info::device::graph_support>)
```
but wasn't aware of the convention to leave things as implementation defined, so will leave the numeric value out

Reply:
Updating to reflect this feedback in https://github.com/reble/llvm/pull/197

Reply:
FWIW, you might consider making this an aspect instead of an "information descriptor" query.  That way, users can write the check like:

```
if (!device::has(ext_oneapi_graph_support))
```

However, this only provides a yes/no answer.  Is it important to distinguish between "native" vs. "emulated"?  Is there even a clear distinction in all cases?  What if the graph contains a host task, for example?  I think you need to split the graph into multiple parts in such a case, right?  Would that be "native" or "emulated"?

Reply:
I think there is a clear distinction, it's if the backend supports a native command-buffer construct we can use or not. We can't accurately reflect this in the spec/documentation as 1) The user might have a different version of the runtime compared to the tip spec, where we might have changed a backend from emulated to native in that period and 2) For the OpenCL backend it is up to the vendor whether they support extensions like cl_khr_command_buffer which we will only know at runtime based on what OpenCL library the ICD loads.

Whether the  "native" vs. "emulated" information is useful, I think it is. We're saying this extension improves performance, but if a user tries the implementation out and doesn't see the wins they were expecting, they may want to know why. Or they may have multiple devices and want to pick the one with "native" rather than "emulated" to use the extension with.

We could possibly have an `aspect` as well, then provide both the benefits of more concise user code using the aspect with extra information from `get_info` if a user cares about those details.